File: tests/algorithms/x/Pascal/maths/addition_without_arithmetic.pas
{$mode objfpc}{$modeswitch nestedprocvars}
program Main;
uses SysUtils;
var _nowSeed: int64 = 0;
var _nowSeeded: boolean = false;
procedure init_now();
var s: string; v: int64;
begin
  s := GetEnvironmentVariable('MOCHI_NOW_SEED');
  if s <> '' then begin
    Val(s, v);
    _nowSeed := v;
    _nowSeeded := true;
  end;
end;
function _now(): integer;
begin
  if _nowSeeded then begin
    _nowSeed := (_nowSeed * 1664525 + 1013904223) mod 2147483647;
    _now := _nowSeed;
  end else begin
    _now := Integer(GetTickCount64()*1000);
  end;
end;
function _bench_now(): int64;
begin
  _bench_now := GetTickCount64()*1000;
end;
function _mem(): int64;
var h: TFPCHeapStatus;
begin
  h := GetFPCHeapStatus;
  _mem := h.CurrHeapUsed;
end;
procedure panic(msg: string);
begin
  writeln(msg);
  halt(1);
end;
procedure error(msg: string);
begin
  panic(msg);
end;
function _to_float(x: integer): real;
begin
  _to_float := x;
end;
function to_float(x: integer): real;
begin
  to_float := _to_float(x);
end;
procedure json(xs: array of real);
var i: integer;
begin
  write('[');
  for i := 0 to High(xs) do begin
    write(xs[i]);
    if i < High(xs) then write(', ');
  end;
  writeln(']');
end;
var
  bench_start_0: integer;
  bench_dur_0: integer;
  bench_mem_0: int64;
  bench_memdiff_0: int64;
  MAX: integer;
  HALF: integer;
  n: integer;
  a: integer;
  b: integer;
  num: integer;
function to_unsigned(n: integer): integer; forward;
function from_unsigned(n: integer): integer; forward;
function bit_and(a: integer; b: integer): integer; forward;
function bit_xor(a: integer; b: integer): integer; forward;
function lshift1(num: integer): integer; forward;
function add(a: integer; b: integer): integer; forward;
function to_unsigned(n: integer): integer;
begin
  if n < 0 then begin
  exit(MAX + n);
end;
  exit(n);
end;
function from_unsigned(n: integer): integer;
begin
  if n >= HALF then begin
  exit(n - MAX);
end;
  exit(n);
end;
function bit_and(a: integer; b: integer): integer;
var
  bit_and_x: integer;
  bit_and_y: integer;
  bit_and_res: integer;
  bit_and_bit: integer;
  bit_and_i: integer;
begin
  bit_and_x := a;
  bit_and_y := b;
  bit_and_res := 0;
  bit_and_bit := 1;
  bit_and_i := 0;
  while bit_and_i < 32 do begin
  if ((bit_and_x mod 2) = 1) and ((bit_and_y mod 2) = 1) then begin
  bit_and_res := bit_and_res + bit_and_bit;
end;
  bit_and_x := bit_and_x div 2;
  bit_and_y := bit_and_y div 2;
  bit_and_bit := bit_and_bit * 2;
  bit_and_i := bit_and_i + 1;
end;
  exit(bit_and_res);
end;
function bit_xor(a: integer; b: integer): integer;
var
  bit_xor_x: integer;
  bit_xor_y: integer;
  bit_xor_res: integer;
  bit_xor_bit: integer;
  bit_xor_i: integer;
  bit_xor_abit: integer;
  bit_xor_bbit: integer;
begin
  bit_xor_x := a;
  bit_xor_y := b;
  bit_xor_res := 0;
  bit_xor_bit := 1;
  bit_xor_i := 0;
  while bit_xor_i < 32 do begin
  bit_xor_abit := bit_xor_x mod 2;
  bit_xor_bbit := bit_xor_y mod 2;
  if ((bit_xor_abit + bit_xor_bbit) mod 2) = 1 then begin
  bit_xor_res := bit_xor_res + bit_xor_bit;
end;
  bit_xor_x := bit_xor_x div 2;
  bit_xor_y := bit_xor_y div 2;
  bit_xor_bit := bit_xor_bit * 2;
  bit_xor_i := bit_xor_i + 1;
end;
  exit(bit_xor_res);
end;
function lshift1(num: integer): integer;
begin
  exit((num * 2) mod MAX);
end;
function add(a: integer; b: integer): integer;
var
  add_first: integer;
  add_second: integer;
  add_carry: integer;
  add_result_: integer;
begin
  add_first := to_unsigned(a);
  add_second := to_unsigned(b);
  while add_second <> 0 do begin
  add_carry := bit_and(add_first, add_second);
  add_first := bit_xor(add_first, add_second);
  add_second := lshift1(add_carry);
end;
  add_result_ := from_unsigned(add_first);
  exit(add_result_);
end;
begin
  init_now();
  bench_mem_0 := _mem();
  bench_start_0 := _bench_now();
  MAX := 4294967296;
  HALF := 2147483648;
  writeln(IntToStr(add(3, 5)));
  writeln(IntToStr(add(13, 5)));
  writeln(IntToStr(add(-7, 2)));
  writeln(IntToStr(add(0, -7)));
  writeln(IntToStr(add(-321, 0)));
  bench_memdiff_0 := _mem() - bench_mem_0;
  bench_dur_0 := (_bench_now() - bench_start_0) div 1000;
  writeln('{');
  writeln(('  "duration_us": ' + IntToStr(bench_dur_0)) + ',');
  writeln(('  "memory_bytes": ' + IntToStr(bench_memdiff_0)) + ',');
  writeln(('  "name": "' + 'main') + '"');
  writeln('}');
end.

File: tests/algorithms/x/Clojure/maths/addition_without_arithmetic.clj
(ns main (:refer-clojure :exclude [to_unsigned from_unsigned bit_and bit_xor lshift1 add]))

(require 'clojure.set)

(defn in [x coll]
  (cond (string? coll) (clojure.string/includes? coll x) (map? coll) (contains? coll x) (sequential? coll) (some (fn [e] (= e x)) coll) :else false))

(defn padStart [s w p]
  (loop [out (str s)] (if (< (count out) w) (recur (str p out)) out)))

(defn indexOf [s sub]
  (let [idx (clojure.string/index-of s sub)] (if (nil? idx) -1 idx)))

(defn split [s sep]
  (clojure.string/split s (re-pattern sep)))

(defn toi [s]
  (Integer/parseInt (str s)))

(defn _fetch [url]
  {:data [{:from "" :intensity {:actual 0 :forecast 0 :index ""} :to ""}]})

(def nowSeed (atom (let [s (System/getenv "MOCHI_NOW_SEED")] (if (and s (not (= s ""))) (Integer/parseInt s) 0))))

(declare to_unsigned from_unsigned bit_and bit_xor lshift1 add)

(def ^:dynamic add_carry nil)

(def ^:dynamic add_result nil)

(def ^:dynamic add_second nil)

(def ^:dynamic bit_and_bit nil)

(def ^:dynamic bit_and_i nil)

(def ^:dynamic bit_and_res nil)

(def ^:dynamic bit_and_x nil)

(def ^:dynamic bit_and_y nil)

(def ^:dynamic bit_xor_abit nil)

(def ^:dynamic bit_xor_bbit nil)

(def ^:dynamic bit_xor_bit nil)

(def ^:dynamic bit_xor_i nil)

(def ^:dynamic bit_xor_res nil)

(def ^:dynamic bit_xor_x nil)

(def ^:dynamic bit_xor_y nil)

(def ^:dynamic first_v nil)

(def ^:dynamic main_MAX nil)

(def ^:dynamic main_HALF nil)

(defn to_unsigned [to_unsigned_n]
  (try (if (< to_unsigned_n 0) (+ main_MAX to_unsigned_n) to_unsigned_n) (catch clojure.lang.ExceptionInfo e (if (= (ex-message e) "return") (get (ex-data e) :v) (throw e)))))

(defn from_unsigned [from_unsigned_n]
  (try (if (>= from_unsigned_n main_HALF) (- from_unsigned_n main_MAX) from_unsigned_n) (catch clojure.lang.ExceptionInfo e (if (= (ex-message e) "return") (get (ex-data e) :v) (throw e)))))

(defn bit_and [bit_and_a bit_and_b]
  (binding [bit_and_bit nil bit_and_i nil bit_and_res nil bit_and_x nil bit_and_y nil] (try (do (set! bit_and_x bit_and_a) (set! bit_and_y bit_and_b) (set! bit_and_res 0) (set! bit_and_bit 1) (set! bit_and_i 0) (while (< bit_and_i 32) (do (when (and (= (mod bit_and_x 2) 1) (= (mod bit_and_y 2) 1)) (set! bit_and_res (+ bit_and_res bit_and_bit))) (set! bit_and_x (/ bit_and_x 2)) (set! bit_and_y (/ bit_and_y 2)) (set! bit_and_bit (* bit_and_bit 2)) (set! bit_and_i (+ bit_and_i 1)))) (throw (ex-info "return" {:v bit_and_res}))) (catch clojure.lang.ExceptionInfo e (if (= (ex-message e) "return") (get (ex-data e) :v) (throw e))))))

(defn bit_xor [bit_xor_a bit_xor_b]
  (binding [bit_xor_abit nil bit_xor_bbit nil bit_xor_bit nil bit_xor_i nil bit_xor_res nil bit_xor_x nil bit_xor_y nil] (try (do (set! bit_xor_x bit_xor_a) (set! bit_xor_y bit_xor_b) (set! bit_xor_res 0) (set! bit_xor_bit 1) (set! bit_xor_i 0) (while (< bit_xor_i 32) (do (set! bit_xor_abit (mod bit_xor_x 2)) (set! bit_xor_bbit (mod bit_xor_y 2)) (when (= (mod (+ bit_xor_abit bit_xor_bbit) 2) 1) (set! bit_xor_res (+ bit_xor_res bit_xor_bit))) (set! bit_xor_x (/ bit_xor_x 2)) (set! bit_xor_y (/ bit_xor_y 2)) (set! bit_xor_bit (* bit_xor_bit 2)) (set! bit_xor_i (+ bit_xor_i 1)))) (throw (ex-info "return" {:v bit_xor_res}))) (catch clojure.lang.ExceptionInfo e (if (= (ex-message e) "return") (get (ex-data e) :v) (throw e))))))

(defn lshift1 [lshift1_num]
  (try (throw (ex-info "return" {:v (mod (* lshift1_num 2) main_MAX)})) (catch clojure.lang.ExceptionInfo e (if (= (ex-message e) "return") (get (ex-data e) :v) (throw e)))))

(defn add [add_a add_b]
  (binding [add_carry nil add_result nil add_second nil first_v nil] (try (do (set! first_v (to_unsigned add_a)) (set! add_second (to_unsigned add_b)) (while (not= add_second 0) (do (set! add_carry (bit_and first_v add_second)) (set! first_v (bit_xor first_v add_second)) (set! add_second (lshift1 add_carry)))) (set! add_result (from_unsigned first_v)) (throw (ex-info "return" {:v add_result}))) (catch clojure.lang.ExceptionInfo e (if (= (ex-message e) "return") (get (ex-data e) :v) (throw e))))))

(defn -main []
  (let [rt (Runtime/getRuntime)
    start-mem (- (.totalMemory rt) (.freeMemory rt))
    start (System/nanoTime)]
      (alter-var-root (var main_MAX) (constantly 4294967296))
      (alter-var-root (var main_HALF) (constantly 2147483648))
      (println (str (add 3 5)))
      (println (str (add 13 5)))
      (println (str (add (- 7) 2)))
      (println (str (add 0 (- 7))))
      (println (str (add (- 321) 0)))
      (System/gc)
      (let [end (System/nanoTime)
        end-mem (- (.totalMemory rt) (.freeMemory rt))
        duration-us (quot (- end start) 1000)
        memory-bytes (Math/abs ^long (- end-mem start-mem))]
        (println (str "{\n  \"duration_us\": " duration-us ",\n  \"memory_bytes\": " memory-bytes ",\n  \"name\": \"main\"\n}"))
      )
    ))

(-main)
